pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
+	if cap(ts) > len(ts) {
+		ts = ts[:newPos+1]
+	} else {
+		ts = append(ts, mimirpb.PreallocTimeseries{})
 	}
 
-	returnErr := func(err error) {
-		if !r.propagateErrors {
-			return
-		}
-		errCh <- err
-	}
+	ts[newPos].TimeSeries = f.pools.getTs()
 
-	var wg sync.WaitGroup
-	wg.Add(len(r.tsByEndpoint))
+	return ts
+}
 
-	// Can't use concurrency.ForEachJob because we don't want to cancel the other jobs if one errors.
-	errorsByEndpoint := make(map[string]error, len(r.tsByEndpoint))
-	var errorsMtx sync.Mutex
-	for endpoint, ts := range r.tsByEndpoint {
-		go func(endpoint string, ts []mimirpb.PreallocTimeseries) {
-			defer wg.Done()
+// worker is a worker go routine which performs the forwarding requests that it receives through a channel.
+func (f *forwarder) worker() {
+	defer f.workerWg.Done()
 
-			errorsMtx.Lock()
-			defer errorsMtx.Unlock()
-			errorsByEndpoint[endpoint] = r.sendToEndpoint(ctx, endpoint, ts)
-		}(endpoint, *ts)
+	for req := range f.reqCh {
+		req.do()
 	}
+}
 
-	go func() {
-		defer r.cleanup()
-		defer close(errCh)
+type request struct {
+	pools  *pools
+	client *http.Client
+	log    log.Logger
 
-		wg.Wait()
+	ctx             context.Context
+	timeout         time.Duration
+	propagateErrors bool
+	errCh           chan error
+	requestWg       *sync.WaitGroup
 
-		var nonRecoverable error
-		// No need to get errorsMtx because we already waited for all routines which might modify it to end.
-		for endpoint, err := range errorsByEndpoint {
-			if err == nil {
-				continue
-			}
+	endpoint string
+	ts       tsWithSampleCount
 
-			if errors.As(err, &recoverableError{}) {
-				// If there is at least one recoverable error we want to return the recoverable error.
-				returnErr(httpgrpc.Errorf(http.StatusInternalServerError, "endpoint %s: %s", endpoint, err.Error()))
-				return
-			}
+	requests  prometheus.Counter
+	errors    *prometheus.CounterVec
+	samples   prometheus.Counter
+	exemplars prometheus.Counter
+	latency   prometheus.Histogram
+}
 
-			nonRecoverable = httpgrpc.Errorf(http.StatusBadRequest, "endpoint %s: %s", endpoint, err.Error())
-		}
+// newRequest launches a new forwarding request and sends it to a worker via a channel.
+// It might block if all the workers are busy.
+func (f *forwarder) newRequest(ctx context.Context, endpoint string, ts tsWithSampleCount, requestWg *sync.WaitGroup, errCh chan error) {
+	req := f.pools.getReq()
+
+	req.pools = f.pools
+	req.client = &f.client // http client should be re-used so open connections get re-used.
+	req.log = f.log
+	req.ctx = ctx
+	req.timeout = f.cfg.RequestTimeout
+	req.propagateErrors = f.cfg.PropagateErrors
+	req.errCh = errCh
+	req.requestWg = requestWg
+
+	// Target endpoint and TimeSeries to forward.
+	req.endpoint = endpoint
+	req.ts = ts
+
+	// Metrics.
+	req.requests = f.requestsTotal
+	req.errors = f.errorsTotal
+	req.samples = f.samplesTotal
+	req.exemplars = f.exemplarsTotal
+	req.latency = f.requestLatencyHistogram
+
+	f.reqCh <- req
```

Comment:
This should respect the `ctx` and stop blocking if context was cancelled.

Reply:
good catch, thx. 
Fixed in f6265e471